### PR TITLE
Gro format rewrite

### PR DESCRIPTION
**Pull request by frodofine**
Closes part of #154.

### CODE REVIEW

File: src/FormatFactory.cpp [hunk 1]
```diff
@@ -45,10 +45,10 @@ FormatFactory::FormatFactory() {
     this->add_format<LAMMPSDataFormat>();
     this->add_format<SDFFormat>();
     this->add_format<CSSRFormat>();
+    this->add_format<GROFormat>();
 
     // VMD molfile plugins
     this->add_format<Molfile<DCD>>();
-    this->add_format<Molfile<GRO>>();
```
Comment (Luthaf) [inline src/FormatFactory.cpp]: Can you also remove the template instantiation in Molfile.cpp ?

File: src/formats/GRO.cpp [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+    size_t natoms = static_cast<size_t>(count);
+
+    residues_.clear();
+    frame.add_velocities();
+    frame.reserve(natoms);
+    frame.resize(0);
+
+    for (const auto& line: file_->readlines(natoms)) {
+        auto resid = std::stoul(line.substr(0,5));
```
Comment (Luthaf) [inline src/formats/GRO.cpp]: could you add a check that the line is long enough?

File: src/formats/GRO.cpp [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        std::string resname = line.substr(5, 5);
+
+        std::string atomname= line.substr(10,5);
+
+        // GRO files store atoms in NM, we need to convert to Angstroms
+        auto x = std::stof(line.substr(20,8)) * 10;
+        auto y = std::stof(line.substr(28,8)) * 10;
+        auto z = std::stof(line.substr(36,8)) * 10;
+
+        if (line.length() > 44) {
```
Comment (Luthaf) [inline src/formats/GRO.cpp]: Same here, could you check that the line is long enough to read all velocity data

File: src/formats/GRO.cpp [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+
+        auto a = std::stof(box_values[0]) * 10;
+        auto b = std::stof(box_values[1]) * 10;
+        auto c = std::stof(box_values[2]) * 10;
+
+        auto cell = UnitCell(a, b, c);
+        frame.set_cell(cell);
+    }
+
+    if (box_values.size() == 9) {
```
Comment (Luthaf) [inline src/formats/GRO.cpp]: Could this be a `else if` clause instead ? Both tests are related

File: src/formats/GRO.cpp [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+        }
+
+    }
+
+    std::string box = file_->readline();
+    auto box_values = split(box, ' ');
+
+    if (box_values.size() == 3) {
+
+        auto a = std::stof(box_values[0]) * 10;
```
Comment (Luthaf) [inline src/formats/GRO.cpp]: Could you use `string2double` from utils.hpp instead of `std::stof`? The former check if the value is indeed a number and throw an error if not.

File: src/formats/GRO.cpp [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+        auto cell = UnitCell(a, b, c);
+        frame.set_cell(cell);
+    }
+
+    if (box_values.size() == 9) {
+        auto v1_x = std::stof(box_values[0]) * 10;
+        auto v2_y = std::stof(box_values[1]) * 10;
+        auto v3_z = std::stof(box_values[2]) * 10;
+
+        assert(std::stof(box_values[3]) == 0);
```
Comment (Luthaf) [inline src/formats/GRO.cpp]: Is this because of the restriction that cell matrix must be upper triangular ?

File: src/formats/GRO.cpp [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+
+    for (auto& residue: residues_) {
+        frame.add_residue(residue.second);
+    }
+}
+
+static std::string to_gro_index(uint64_t i) {
+    auto id = i + 1;
+
+    if (id >= 100000) {
```
Comment (Luthaf) [inline src/formats/GRO.cpp]: you could just check `if (i + 1 >= 100000)`

File: src/formats/GRO.cpp [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+        warning("Too many atoms for GRO format, removing atomic id");
+        return "*****";
+    } else {
+        return std::to_string(i + 1);
+    }
+}
+
+void GROFormat::write(const Frame& frame) {
+    if (frame.get("name")) {
+        fmt::print(*file_, frame.get("name")->as_string() + "\n");
```
Comment (Luthaf) [inline src/formats/GRO.cpp]: Do we want to check that `frame.get("name")` is indeed a string property ?

File: src/formats/GRO.cpp (hunk 6 above)
Comment (frodofine) [reply src/formats/GRO.cpp]: No, it comes from the format itself. For some reason there must be zeros in those locations, see [this](http://manual.gromacs.org/current/online/gro.html):

> values: v1(x) v2(y) v3(z) v1(y) v1(z) v2(x) v2(z) v3(x) v3(y), the last 6 values may be omitted (they will be set to zero). Gromacs only supports boxes with v1(y)=v1(z)=v2(z)=0. >

File: src/formats/GRO.cpp (hunk 2 above)
Comment (frodofine) [reply src/formats/GRO.cpp]: Fair enough!

File: src/formats/GRO.cpp (hunk 3 above)
Comment (frodofine) [reply src/formats/GRO.cpp]: Meant to use 68, thanks!

File: src/formats/GRO.cpp (hunk 5 above)
Comment (frodofine) [reply src/formats/GRO.cpp]: Sure thing.

File: src/formats/GRO.cpp (hunk 7 above)
Comment (frodofine) [reply src/formats/GRO.cpp]: Sure

File: src/formats/GRO.cpp (hunk 6 above)
Comment (Luthaf) [reply src/formats/GRO.cpp]: Ok, thanks!

File: src/formats/GRO.cpp (hunk 8 above)
Comment (frodofine) [reply src/formats/GRO.cpp]: Sure

File: .travis.yml [hunk 9]
```diff
@@ -81,7 +81,7 @@ script:
   - mkdir build && cd build
   - $CMAKE_CONFIGURE cmake $CMAKE_ARGS $CMAKE_EXTRA ..
   - make -j2
-  - ctest -j2 --output-on-failure
+  - travis_wait 30 ctest -j2 --output-on-failure
```
Comment (Luthaf) [inline .travis.yml]: why is this needed ?

Comment (frodofine) [reply .travis.yml]: Travis during the Valgrind test is timing out because no output is received for 10 minutes since the PDB and GRO tests take longer than this to complete (and they take up both cores). This is to stop Travis from killing the job. I'm happy to look into other fixes for this. For example, is `--track-origins=yes` needed? I know this can increase Valgrind runtime significantly.

On my machine, both PDB and GRO tests complete in seconds, not 10 minutes - so I'm not sure why Travis takes so long.

Comment (Luthaf) [reply .travis.yml]: Ok! I did not know what the travis_wait command did :smile: 

`--track-origins=yes` might not be needed here, and I have a local branch to improve this a bit too by not running the longest test with Travis + valgrind. I'll push it as soon as possible.

> On my machine, both PDB and GRO tests complete in seconds, not 10 minutes - so I'm not sure why Travis takes so long.

By default, the tests are not run under valgrind. And Travis machines are not particularly powerful ones.

Comment (frodofine) [reply .travis.yml]: I meant with `-DVALGRIND=ON` used on my local machine :smiley: 

I think the issue is that Travis does not provide powerful machines, but its free so I can't complain :roll_eyes: .

`travis_wait 30` seems to have solved the issue (looking at the timing I could have used a number smaller than 30, I just wanted to be sure).

File: src/FormatFactory.cpp (hunk 1 above)
Comment (frodofine) [reply src/FormatFactory.cpp]: Whoops forgot about this.
